apollo-router/src/plugins/telemetry/mod.rs

```diff
@@ -1245,6 +1251,58 @@ impl Telemetry {
         };
         sender.send(SingleReport::Stats(metrics));
     }
+
+    fn decode_subgraph_ftv1(context: &Context) -> (u64, HashMap<String, SingleTypeStat>) {
+        use crate::plugins::telemetry::apollo_exporter::proto::reports::trace::Node;
+
+        fn traverse(
+            per_type: &mut HashMap<String, SingleTypeStat>,
+            field_execution_weight: f64,
+            node: &Node,
+        ) {
+            let single_field_stat = per_type
+                .entry(node.parent_type.clone())
+                .or_default()
+                .per_field_stat
+                .entry(node.original_field_name.clone())
+                .or_insert_with(|| metrics::apollo::studio::SingleFieldStat {
+                    return_type: node.r#type.clone(), // not `Default::default()`’s empty string
+                    errors_count: 0,
+                    latency: Default::default(),
+                    estimated_execution_count: 0.0,
+                    requests_with_errors_count: 0,
+                });
+            single_field_stat.errors_count += node.error.len() as u64;
+            single_field_stat.latency.increment_duration(
+                Duration::from_nanos(node.end_time.saturating_sub(node.start_time)),
+                field_execution_weight,
+            );
+            single_field_stat.estimated_execution_count += field_execution_weight;
+            if !node.error.is_empty() {
+                single_field_stat.requests_with_errors_count += 1;
+            }
+            for child in &node.child {
+                traverse(per_type, field_execution_weight, child)
+            }
+        }
+
+        let mut requests_without_field_instrumentation = 0;
+        let mut per_type = HashMap::new();
+
+        if let Some(Value::Array(array)) = context.get_json_value(SUBGRAPH_FTV1) {
+            for trace in array
+                .iter()
+                .filter_map(|value| decode_ftv1_trace(value.as_str()?))
+            {
+                if trace.field_execution_weight == 0.0 {
```

Comment:
Is this looking for a `field_execution_weight` on the ftv1 trace sent by the subgraph? I don't think that's a thing. (You can see in reports.proto that that field is below the "Fields below this line are *not* included in federated traces" comment.) At least in AS/AG, the FEW is just calculated by the gateway server, not by individual subgraphs. Does this work?

Reply:
Ie you should calculate this yourself based on your sampling fraction

Reply:
This is porting https://github.com/apollographql/apollo-server/blob/fa82c1d5299/packages/server/src/plugin/usageReporting/stats.ts#L278 which I’m told is Gateway code (despite being in the apollo-server repo). Is it not?

Reply:
It's gateway code in the sense that gateway servers run both Apollo Server and Apollo Gateway combined, and that this is thus the code that sends reports from Gateway. But this function is called once per overall (monolith/supergraph) operation.

(btw are you actually basing this off off of fa82c1d5299? that's rather old...)

Reply:
Alright, after looking at this again with Bryn we’ve figured out that that `addTrace` is called with a single trace for the entire supergraph request, which includes nested traces from subgraphs somewhere under its `queryPlan` field. `iterateOverTrace` accounts for these nested traces too, but the `fieldExecutionWeight` taken from the top-level trace is set by the Gateway itself, it’s not from subgraph traces.

That commit link was from https://github.com/mdg-private/planning/issues/1814#issuecomment-1413390038, maybe it’s the one that GitHub search happened to index? Anyway I’ve changed the code comment in the PR to https://github.com/apollographql/apollo-server/blob/6ff88e87c52/packages/server/src/plugin/usageReporting/stats.ts#L283, which doesn’t look too different.